### PR TITLE
Implement binomial distribution

### PR DESCRIPTION
**Pull request by alicanb**
@apaszke can you take a look?

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    Args:
+        total_count (int): number of trials
+        probs (Tensor or Variable): event probabilities
+        logits (Tensor or Variable): event log probabilities
+    """
+    params = {'probs': constraints.unit_interval}
+    has_enumerate_support = True
+
+    def __init__(self, total_count=1, probs=None, logits=None):
```
Comment (apaszke) [inline torch/distributions/binomial.py]: Weren't we using log-odds for `logits` of `Bernoulli`? Is this consistent with TensorFlow?

File: torch/distributions/utils.py [hunk 2]
```diff
@@ -132,15 +132,24 @@ def logits_to_probs(logits, is_binary=False):
...
 
 
+def clamp_probs(probs):
+    eps = _finfo(probs).eps
+    return probs.clamp(min=eps, max=1 - eps)
+
+
+def clamp_logits(logits):
+    eps = _finfo(logits).eps
+    return logits.clamp(min=math.log(eps), max=math.log1p(-eps))
```
Comment (apaszke) [inline torch/distributions/utils.py]: Have you checked that `log`/`log1p` are accurate in here when called with `eps`?

Comment (alicanb) [reply torch/distributions/utils.py]: No, but I wrote that just for completeness. I removed it to be safe

File: torch/distributions/binomial.py (hunk 1 above)
Comment (alicanb) [reply torch/distributions/binomial.py]: Yep, my bad, just fixed it.

**Comment by alicanb** [comment]
That fail doesn't seem to be related with this PR

File: torch/distributions/binomial.py (hunk 1 above)
Comment (apaszke) [reply torch/distributions/binomial.py]: @fritzo this is exactly why I wanted to use a different name. Calling two semantically different things the same is going to end up in an error sooner or later without a type system

File: torch/distributions/binomial.py [hunk 3]
```diff
@@ -75,7 +75,7 @@ def log_prob(self, value):
         log_factorial_k = torch.lgamma(value + 1)
         log_factorial_nmk = torch.lgamma(self.total_count - value + 1)
         return (log_factorial_n - log_factorial_k - log_factorial_nmk +
-                value * self.logits + (self.total_count - value) * torch.log1p(-probs))
+                value * self.logits + self.total_count * torch.log1p(-probs))
```
Comment (apaszke) [inline torch/distributions/binomial.py]: Has someone reviewed this math? New version isn't equivalent to the old one

Comment (alicanb) [reply torch/distributions/binomial.py]: I used to use `logits` here as `probs.log()`, now I fixed it to be log odds which is `probs.log() - torch.log1p(-probs))` , so they are equivalent. essentially, `value * old_logits - value * torch.log1p(-probs)) + self.total_count * torch.log1p(-probs))`  = `value * new_logits + self.total_count * torch.log1p(-probs))`

File: torch/distributions/binomial.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/binomial.py]: Fair enough. Should we switch to `log_probs` and `log_odds` then?

File: test/test_distributions.py [hunk 4]
```diff
@@ -298,6 +302,38 @@ def test_bernoulli_3d(self):
                          (2, 5, 2, 3, 5))
         self.assertEqual(Bernoulli(p).sample_n(2).size(), (2, 2, 3, 5))
 
+    def test_binomial(self):
+        total_count = 10
+        p = Variable(torch.Tensor([0.1, 0.2, 0.3]), requires_grad=True)
+        self._gradcheck_log_prob(lambda p: Multinomial(total_count, p), [p])
```
Comment (fritzo) [inline test/test_distributions.py]: Do you want to use `Binomial` rather than `Multinomial` here?

File: test/test_distributions.py [hunk 5]
```diff
@@ -298,6 +302,38 @@ def test_bernoulli_3d(self):
...
+        self._gradcheck_log_prob(lambda p: Multinomial(total_count, None, p.log()), [p])
+        self.assertRaises(NotImplementedError, Multinomial(10, p).rsample)
+        self.assertRaises(NotImplementedError, Multinomial(10, p).entropy)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_binomial_log_prob(self):
+        total_count = 10
+        probs = Variable(torch.Tensor([0.1, 0.2, 0.3]), requires_grad=True)
```
Comment (fritzo) [inline test/test_distributions.py]: Could you add some more test data to provide more assurance that `log_prob` is correct?

File: torch/distributions/binomial.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/binomial.py]: @alicanb I trust tests more than manual inspection of code. Could you add some more test data in `test_binomial_log_prob()`, something like?
```py
probs = torch.arange(0.01, 1, 0.1) 
```
and check multiple total sizes `[0, 1, 2, 10]`?

**Comment by fritzo** [comment]
@apaszke I've added more test data for Alican, at his request.

**Comment by fritzo** [comment]
> @apaszke this [logits ambiguity] is exactly why I wanted to use a different name. Calling two semantically different things the same is going to end up in an error sooner or later without a type system

I hear you and I agree that `probs_to_logits(..., binary=False)` smells bad. I still feel like it's worth maintaining compatibility with Tensorflow arguments despite their loose naming.

If do want to change, I think the minimal path would be to change `Categorical`, `OneHotCategorical`, and `Multinomial` to use `logprobs`. We might go further and change `Bernoulli` and `Binomial` to use `logodds`. Let me know what you decide and I'll volunteer to send you a PR.

cc @neerajprad

**Comment by neerajprad** [comment]
Regarding the naming, I think `logits` is general enough to describe a mapping of values from [0, 1] to [-inf, inf]. For the multi-class case, this mapping is the log of unnormalized probability values (as such, even `logprobs` is more precise than what the mapping actually is). For Bernoulli, we could use the same mapping as the multi-class case, if we were using two parameters, and the naming wouldn't be an issue. But since we use just one parameter, we have to use log odds to map from [0, 1] to [-inf, inf]. It will be nice to keep it consistent with TensorFlow's nomenclature, but if we decide to change it, an alternate naming scheme could be `logits` for Bernoulli and `multinomial_logits` (or, `multinom_logits`) for the multi-class case, so as to cause the least amount of surprise to users used to TensorFlow's API.

**Comment by apaszke** [comment]
I'm not concerned about the naming itself, and I'll leave the final decision up to you. I think changing the names in a consistent way isn't terrible, especially that it requires you to describe your intent more explicitly and we can keep the names similar. I'm worried that if it's easy enough to miss for developers of `torch.distributions`, then some users will trip over it for sure.

**Comment by alicanb** [comment]
I think if we keep the docs consistent (notice how it says log probabilities instead of log-odds here, I'm gonna fix that in a sec. ) we can use `logits` everywhere. If not, maybe `multiclass_logits` or `multi_logits` is a better name since it will also be used in Categorical.

File: torch/distributions/kl.py [hunk 6]
```diff
@@ -157,6 +158,15 @@ def _kl_bernoulli_bernoulli(p, q):
     return t1 + t2
 
 
+@register_kl(Binomial, Binomial)
+def _kl_binomial_binomial(p, q):
+    # from https://math.stackexchange.com/questions/2214993/
+    # kullback-leibler-divergence-for-binomial-distributions-p-and-q
+    if p.total_count != q.total_count:
+        raise(NotImplementedError, 'KL between Binomials with different total_count is not implemented')
```
Comment (apaszke) [inline torch/distributions/kl.py]: That's probably not what you wanted to do. You're raising a tuple as an exception

**Comment by alicanb** [comment]
@neerajprad is sending a PR that fixes the lazy evaulation, @apaszke can you merge that first?

**Review by apaszke — APPROVED** [review]
